charger/fbdect.go

```diff
@@ -0,0 +1,353 @@
+package charger
+
+import (
+	"crypto/md5"
+	"encoding/binary"
+	"encoding/hex"
+	"encoding/xml"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+	"unicode/utf16"
+	"unicode/utf8"
+
+	"github.com/andig/evcc/api"
+	"github.com/andig/evcc/util"
+	"github.com/andig/evcc/util/request"
+)
+
+// AVM FritzBox AHA interface and authentification specifications:
+// https://avm.de/fileadmin/user_upload/Global/Service/Schnittstellen/AHA-HTTP-Interface.pdf
+// https://avm.de/fileadmin/user_upload/Global/Service/Schnittstellen/AVM_Technical_Note_-_Session_ID.pdf
+
+// fbdect AHA-HTTP StatusResponse is the API response if status not OK
+type fbdectStatusResponse struct {
+	Name        string  // DECT Switch name
+	Present     int64   // 0/1 - DECT Switch connected to fritzbox (no/yes)
+	State       int64   // 0/1 - DECT Switch state off/on (empty if unkown or error)
+	Temperature float64 // Wert in 0,1 °C, negative und positive Werte möglich
+	Power       float64 // Wert in 0,001 W (aktuelle Leistung, wird etwa alle 2 Minuten aktualisiert)
+	Energy      float64 // Wert in 1.0 Wh (absoluter Verbrauch seit Inbetriebnahme)
+}
+
+// FbDect charger implementation
+type FbDect struct {
+	*request.Helper
+	uri, ain, user, password, sid string
+	updated                       time.Time
+}
+
+func init() {
+	registry.Add("fbdect", NewFbDectFromConfig)
+}
+
+// NewFbDectFromConfig creates a fbdect charger from generic config
+func NewFbDectFromConfig(other map[string]interface{}) (api.Charger, error) {
+	cc := struct {
+		URI      string
+		AIN      string
+		User     string
+		Password string
+		SID      string
+		Cache    time.Duration
+	}{
+		URI:      "",
+		AIN:      "",
+		User:     "",
+		Password: "",
+		SID:      "",
+		Cache:    0,
+	}
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	if cc.URI == "" || cc.AIN == "" {
+		return nil, errors.New("fbdect config: must have uri and ain of AVM FritzDECT switch")
+	}
+
+	return NewFbDect(cc.URI, cc.AIN, cc.User, cc.Password, cc.SID, cc.Cache)
+}
+
+// NewFbDect creates FbDect charger
+func NewFbDect(uri, ain, user, password, sid string, cache time.Duration) (*FbDect, error) {
+	c := &FbDect{
+		Helper:   request.NewHelper(util.NewLogger("fbdect")),
+		uri:      strings.TrimRight(uri, "/"),
+		ain:      strings.TrimSpace(ain),
+		user:     strings.TrimSpace(user),
+		password: strings.TrimSpace(password),
+		sid:      strings.TrimSpace(sid),
+	}
+
+	return c, nil
+}
+
+func (c *FbDect) fbResponse(function string) string {
+
+	loginURL := fmt.Sprintf("%s/webservices/homeautoswitch.lua", c.uri)
+	parameters := make(map[string]string)
+	// Refresh Fritzbox session id
+	if time.Since(c.updated).Minutes() >= 10 {
```

Comment:
>= 10*time.Minute?

Reply:
Don't get this hint. We have to do a session id refresh after 10 minutes.